### DISCUSSION TITLE
Raylib examples?

### DISCUSSION THREAD

**Discussion by Shadowblitz16** [AvaloniaUI/Avalonia#20416] [Q&A]
### Is your feature request related to a problem? Please describe.

Raylib is a c wrapper library for opengl there is a port for C# called raylib-cs
It would be cool if you could provide and example to hook into raylib's main window and provide texture and renderTexture widgets.

Not sure what is needed to do this but it would be really cool.

### Describe the solution you'd like

Examples for hooking raylib-cs's MainWindow, RenderTexture2D and Texture2D into custom avalonia widgets.

### Describe alternatives you've considered

I have considered using ImGui but it's unsafe and buggy.
I have also tried writing a my own opengl wrapper in C# but it's not easy.

### Additional context

Raylib's uses a static window api so there can only be one window.

**Comment by stevemonaco** [comment]
Sharing responsibility between multiple renderers is difficult. There are several scenarios here and it's not clear what you want:

1. Have an Avalonia app and want to embed a window's content drawn by Raylib as a child view -> use `NativeControlHost` or integrate with the compositor.
2. Have a Raylib app and want to use Avalonia -> Create a new Avalonia platform, probably.
3. Want to use Raylib to render Avalonia views -> create a vector rendering engine in Raylib and create a new Avalonia rendering backend.

Some are easier than others and mixed content tends to run into airspace issues.

**Comment by Shadowblitz16** [comment]
I want to render the raylib window inside Avalonia and have Avalonia controls that render raylib render textures and textures

**Comment by stevemonaco** [comment]
Shouldn't be that difficult to render the Raylib window by wrapping it with `NativeControlHost`. Rendering Raylib textures as an Avalonia control will be much harder. AFAIK, any approach involving `WriteableBitmap` will pull from the GPU (previously rendered Raylib texture) to CPU (`WriteableBitmap` locked and being mutated with texture data), then push back to GPU to draw. I imagine there's a compositor-based approach for this, but the entire area is largely undocumented. Said implementation would be up to you. `OpenGlControlBase` might provide some inspiration.

**Comment by VadikPtr** [comment]
I am trying to do similar thing (not with raylib, but with my own rendering/platform framework). And as far as I tried, there is no way to implement custom platform for avalonia outside of avalonia. I don't understand why everything I can implement myself is made "internal" and the only way is to download avalonia sources and patch it.

**Comment by stevemonaco** [comment]
@cprkv Custom backends are supported though the APIs have no version stability guarantees. You need to opt-in using `<AvaloniaAccessUnstablePrivateApis>true</AvaloniaAccessUnstablePrivateApis>` in csproj and then you will be able to work with `[PrivateApi]`, `Impl`-ending interfaces, etc. This exists so ordinary applications don't use these APIs and unknowingly take on higher breaking change risks and also so Avalonia has an easier time being split into multiple packages.
